Lower time required for heading recovery after very hard (involontary) landings

tbolin opened the pull request:
Work in progress for several small improvements to speed up the recovery after crashing.

## Improvements so far
###  Avoid decreasing accelerometer gain at large errors
The error between the estimated up direction and the upward direction measured by the accelerometer is calculated by a cross product. Both vectors have magnitude 1, so the norm of the error is just the absolute value of sin of the angle between the vectors. For angles with an absolute value of more than 90 degrees the magnitude of the error would decrease, slowing down convergence. This commit introduces a check for those large angles and ensure that the magnitude of the error always increases when the angle increases.
Similar to https://github.com/betaflight/betaflight/pull/12792 but for the accelerometer.

## Planned improvements
* Accelerometer trust bases on difference from 1g and current rotational velocity (currently yes/no for velocity and not accounting for rotation)
* Calculate accelerometer gain based on reliability of current estimate.

blckmn commented:
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

haslinghuis commented:
@tbolin please rebase on master.

Please explain the tent function :)

tbolin commented:
> Please explain the tent function :)
https://en.wikipedia.org/wiki/Triangular_function
It can among other things be used as a rough but fast approximation of a gaussian.
I chose to call it *tent* since it's also a common name and less likely to conflict with something else than *triangle* or *triangular*.

I'll rebase tomorrow.

haslinghuis commented:
:rofl: Yes was looking at:

https://en.wikipedia.org/wiki/Tent_map
https://math.stackexchange.com/questions/4669545/help-understanding-discription-of-a-mathematical-tent

tbolin commented:
Rebased on master and fixed an issue with building the SITL target.
The tests failing is expected, and I will get to that ones I have managed to get the test target to build locally.

tbolin commented on src/main/common/maths.c:
```diff
@@ -134,6 +134,15 @@ float fapplyDeadband(const float value, const float deadband)
     return value >= 0 ? value - deadband : value + deadband;
 }
 
+float tent(const float x, const float center, const float width)
```
I'll add comments and a wiki link.
Center probably isn't necessary, I will remove it.

tbolin replied:
Comment with wiki link added and center removed

tbolin commented on src/main/flight/imu.c:
```diff
@@ -290,10 +291,23 @@ STATIC_UNIT_TESTED void imuMahonyAHRSupdate(float dt, float gx, float gy, float
         ay *= recipAccNorm;
         az *= recipAccNorm;
 
+        fpVector3_t acc_upp = {.x = ax, .y = ay, .z = az};
```
I'll change the suffixes. What does ef and bf stand for here?
And when I think about it it should probably be `AccEF` and `AccBF` to comply with Betaflight's coding standards.

ledvinap replied:
BodyFrame and EarthFrame. Rest of code uses underscore suffix, but using camel case ma be good idea. (personally, I slightly prefer underscore version in this case (reference frame is separated from variable name)

tbolin replied:
*upp* should be *up* so `acc_up` means the direction of up as measured by the accelerometer. In the same way `est_up(p)` is the previous estimate of the upwards direction.
Unless I'm missing something both vectors are in the body frame here, and should probably have a bf suffix in addition to the direction.
I'll keep the snake case suffixes for now.

tbolin commented on src/main/flight/imu.c:
```diff
@@ -489,10 +459,19 @@ static void imuComputeQuaternionFromRPY(quaternionProducts *quatProd, int16_t in
+static float imuUpdateDeviation(const float deviation, const float imuDt, const float durationSaturated)
+{
+    // Increase estiumated deviation based on time spent integrating
+    static const float err_rate_saturated = 500.0f; // guestimated gyro drift in deg/s when saturated
```
I'm not sure what you are referring to here. use selected gyro range instead of what?

tbolin commented on src/main/flight/imu.c:
```diff
@@ -366,80 +380,36 @@ STATIC_UNIT_TESTED void imuUpdateEulerAngles(void)
+
+    // decrease the estimated deviation based on how much we currently trust the acc
+    static const float heuristic = 0.1f;
+    const float omega = 2.0f * M_PIf * dt * (heuristic * kpGain);
```
The short rationale is that since the filter should converge exponentially the trust in the current estimate should also converge exponentially. 
It's intended to use basically the same math as the PT1 filter, but I was quite tired when I wrote that code.
I've also changed the code to use the base kp gain now, since I had problem with the deviation converging to quickly during flight tests.
I will definitely revisit this part.

tbolin commented on src/main/flight/imu.c:
```diff
@@ -489,10 +459,19 @@ static void imuComputeQuaternionFromRPY(quaternionProducts *quatProd, int16_t in
+{
+    // Increase estiumated deviation based on time spent integrating
+    static const float err_rate_saturated = 500.0f; // guestimated gyro drift in deg/s when saturated
+    static const float err_rate_normal = 5.0f;  // static gyro drift in deg/s under normal circumstances
```
They should be much better yes, at least after the short calibration betaflight does at power on.
However, I had problems with drift when wobbling the quad by hand, and increasing the normal error rate made them converge faster.
After the flight testing adjustments I have now lowered this rate to 1.0, which is a bit more sensible and should be about what could be expected in a worst case scenario for a normal IMU.

tbolin commented on src/main/flight/imu.c:
```diff
@@ -366,80 +380,36 @@ STATIC_UNIT_TESTED void imuUpdateEulerAngles(void)
 }
 
-static bool imuIsAccelerometerHealthy(float *accAverage)
+static float imuAccTrust(const float* accAverage, const float* gyroAverage, const float gRecip)
```
Agreed on both accounts. I'll rename the variable, and take a look at moving the scaling outside.

tbolin commented on src/main/flight/imu.c:
```diff
@@ -576,7 +564,7 @@ static int calculateThrottleAngleCorrection(void)
 void imuUpdateAttitude(timeUs_t currentTimeUs)
 {
-    if (sensors(SENSOR_ACC) && acc.isAccelUpdatedAtLeastOnce) {
+    if (sensors(SENSOR_ACC) && acc.isAccelUpdatedAtLeastOnce && gyroIsCalibrationComplete()) {
```
I thought so too. However I had problems where the estimated attitude would end up way off if the quad was moved when connecting the power source and for perhaps 5 seconds after. It's almost as if there are some huge jump in the gyro data during or after calibration.
Adding this check removed the issue.
Might be worth looking into further, but currently this check is needed.

tbolin replied:
The gyroADC and gyroADCf values won't be populated unless `gyroCalibrationCompleted` is True, see https://github.com/betaflight/betaflight/blob/d5d3ee88beb63ded0f9c632c259cc31b0e9ad58e/src/main/sensors/gyro.c#L417
The estimate covariance is already initiated to the maximum value at start, and converges quickly once it's getting data. However, the covariance should probably reset each time `gyroCalibrationCompleted` is false, to avoid weirdness if another calibration is initiated somehow.

ledvinap replied:
(Or simply use gyro data during calibration ... )

tbolin replied:
As I wrote before:
> The gyroADC and gyroADCf values won't be populated unless gyroCalibrationCompleted is True

Using uncalibrated data would be fine, using zeroed data will cause issues. Once the calibration is completed the filter converges so fast that it doesn't really matter anyways.

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -290,10 +291,23 @@ STATIC_UNIT_TESTED void imuMahonyAHRSupdate(float dt, float gx, float gy, float
+        // To avoid the gain decreasing for angles > 90 degrees:
+        // set magnitude of error vector to 1 + |cos| of angle between estimated
+        // and measured downwards vectors if the absolute angle of the error is > 90 degrees
+        acc_err = (dot > 0) ? acc_err : *vectorScale(&acc_err, vectorNormalize(&acc_err, &acc_err), 1.0f - dot);
```
I'd prefer simpler version, but it's your call

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -312,9 +326,9 @@ STATIC_UNIT_TESTED void imuMahonyAHRSupdate(float dt, float gx, float gy, float
     }
 
     // Apply proportional and integral feedback
-    gx += dcmKpGain * ex + integralFBx;
```
At least for now, I'd use individual gains for error source (magnetometer/cog/acc), with value close to 1, and then dcmKpGain related to stabilizing PI controller.

Also, applying dcmKpGain early affects dcmKiGain (but common dcmKiGain is more problematic, it is used for offset estimation and should probably be very different (or zero) for CoG / yaw estimation)

tbolin replied:
I'm aware that the Ki currently isn't handled. I noticed it was 0 by default and left it alone for now. Something will have to be done about it later.
I think I will change the  code so the dcmKpGain setting is used in the end of the mahoney update and replace the argument to the mahoney function with the gain for the accelerometer. That will at least focus the influence of the dcmKpGain to one point in the code and not spread it all over.

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -332,6 +348,63 @@ STATIC_UNIT_TESTED void imuMahonyAHRSupdate(float dt, float gx, float gy, float
     q.z += (+buffer.w * gz + buffer.x * gy - buffer.y * gx);
 
+#else
+    const quaternion k1 = {
```
(this is IMO overkill. You will gain more by constructing rotation quaternion properly (cos(alpha)*[w], sin(alpha)*[x,y,z]), see iNav implementation, possibly using low-order approximation for sin/cos)

tbolin replied:
It is overkill. I used it to see if the current integration method might have been the source of the problem I observed during the wobble tests. The integrations doesn't appear to be the source of the drift so this will be removed.

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -366,80 +380,36 @@ STATIC_UNIT_TESTED void imuUpdateEulerAngles(void)
-    }
+    const float deviationGainMax = 100.0f * baseKp;
+    const float deviationGainMin = 0.1f * baseKp;
+    const float kpGain = scaleRangef(*deviation, 0.0f, 180.0f, deviationGainMin, deviationGainMax) * accTrust;
```
source.

Also, using linear scaling to 3 orders of magnitude range feels wrong

tbolin replied:
The deviation is (at least intended) to be in degrees, e.g. if it's 180 the algorithm assumes that the current estimate could be up to 180 degrees off in any direction. The deviation have to be in degrees (or perhaps radians) for the deviation update to make any sense, so scaling the deviation to 0..1 would only create an intermediate scaling step.

The range feels a bit large, yes, but if the deviation is 180 degrees, the current estimate is completely unreliable, so the next estimate should almost be identical to whatever the accelerometer have measured, so the maximum needs to be quite large.
The current upper bound of 25 is taken from the old value of  `ATTITUDE_RESET_KP_GAIN`, so it's not just taken out of the air.

tbolin replied:
Yes, I've come to the same conclusion. The lower bound was originally kpGain and I think I introduced it either before the gyro drift or when the normal gyro drift was very small. Setting it to 0 makes a lot more sense now.

tbolin commented on src/main/flight/imu.c:
```diff
@@ -290,10 +291,23 @@ STATIC_UNIT_TESTED void imuMahonyAHRSupdate(float dt, float gx, float gy, float
         az *= recipAccNorm;
 
+        fpVector3_t acc_upp = {.x = ax, .y = ay, .z = az};
+        fpVector3_t est_upp = {.x = rMat[2][0], .y = rMat[2][1], .z = rMat[2][2]};
```
If the accelerometer is at rest the measured acceleration will point up, in the opposite direction of gravitation.
I agree with the statement about the code being hacked together and needing a bit of cleanup though.

ledvinap commented:
@tbolin: Alternate approach may be to use principles from Kalman filter (yes, I know). Estimate (or more likely guess) noise value for accelerometer (expected noise in down vector) / gyro (saturated/calibrating/normal) and then keep track of attitude estimate noise, as single value. The math will be very similar to this PR, but with cleaner mathematical background.

tbolin commented:
> @tbolin: Alternate approach may be to use principles from Kalman filter (yes, I know). Estimate (or more likely guess) noise value for accelerometer (expected noise in down vector) / gyro (saturated/calibrating/normal) and then keep track of attitude estimate noise, as single value. The math will be very similar to this PR, but with cleaner mathematical background.

I've worked a little bit with Kalman filters before, and I will take a look at if anything in this PR can be brought in line with that. However, I think the biggest problem will be how to translate the various noises into something that can be used in the Mahoney filter to make it converge in at a rate appropriate for how fast the estimate noise converges. This is also a core problem with the current PR. There really isn't a well reasoned link between the convergence rates of the attitude estimate and the attitude deviation estimate.

ledvinap commented:
> However, I think the biggest problem will be how to translate the various noises into something that can be used in the Mahoney filter to make it converge in at a rate appropriate for how fast the estimate noise converges

Mahony is in fact complementary filter (or close to one. It is described/implemented as PI regulator, but math behind it is  equivalent). So Kp is (almost) equivalent to complementary filter gain. And if I understand Kalman correctly, it is calculating optimum gains based on measurement noise and state noise (and updating state noise). So this should provide gain quite easily.
And input noises are mostly guessed / tuned for  given application, and in effect multiply computed gains. So it may be enough only to use correct structure.

tbolin commented:
> > However, I think the biggest problem will be how to translate the various noises into something that can be used in the Mahoney filter to make it converge in at a rate appropriate for how fast the estimate noise converges
> 
> Mahony is in fact complementary filter (or close to one. It is described/implemented as PI regulator, but math behind it is equivalent). So Kp is (almost) equivalent to complementary filter gain. And if I understand Kalman correctly, it is calculating optimum gains based on measurement noise and state noise (and updating state noise). So this should provide gain quite easily. And input noises are mostly guessed / tuned for given application, and in effect multiply computed gains. So it may be enough only to use correct structure.

In the complementary filters I've seen the gain has been in the range [0, 1], and used for a linear interpolation between the state given by the two sensors e. g. $x_{t+1} = (1 - k)a + kb.$ So a gain of 0 means the integrated values are completely trusted, and a gain of 1 means that the state instantly goes to the latest measurement.
IIRC The Kalman gain in a scalar filter is used the same way and is in the same range (with more states I guess the singular values of the gain matrix are in the same range, but I haven't looked that up).
In the code the gain is just $k > 0,$ and as far as I can tell it's the same in the paper that describes the Mahoney filter.
If there exists a k s.t. the state of the filter instantly goes to the latest acc measurement, then that gain could be used as the end of the range instead of 1, essentially implementing a dead beat controller. However, that value  probably is dependent on the error, and even if we find a formula calculating the correct value is likely relatively expensive.

ledvinap commented:
> In the complementary filters I've seen the gain has been in the range [0, 1], and used for a linear interpolation between the state given by the two sensors e. g. xt+1=(1−k)a+kb. So a gain of 0 means the integrated values are completely trusted, and a gain of 1 means that the state instantly goes to the latest measurement.

You can rewrite it as (with 
$x_{t+1} = (1 - k)x_t + kb$ 
$x_{t+1} = x_t + k(b - x_t)$ 

Then, in Mahony, $b - x_t$ (error estimate) is calculated directly from different error sources and $k$ is `dcm_kp`. (It is also similar to RC filter, which is useful to estimate Mahony time constants). `dcm_kp` is set to be quite low ($k\ll1$), so nobody probably considered case where k is close to (or over) one.


> In the code the gain is just k>0, and as far as I can tell it's the same in the paper that describes the Mahoney filter. If there exists a k s.t. the state of the filter instantly goes to the latest acc measurement, then that gain could be used as the end of the range instead of 1, essentially implementing a dead beat controller. However, that value probably is dependent on the error, and even if we find a formula calculating the correct value is likely relatively expensive.

Quaternion (rotation in general) and nonlinear, but most of Mahony (at least in BF) is using first-order approximations (or simply operations that point in correct direction), so setting $k=1$ will be very problematic. But in reality, immediate convergence is not necessary, running few tens of iterations (at 500Hz) is perfectly fine. So gain should be limited to some reasonable value (even limiting time constant using equations for RC filter would be quite easy).

Nice thing about Kalman [based derivation] is that tuning will work with noise magnitudes. The have some  physical meaning, there are theories how to estimate them (blackbox logs are available). Actual gain calculated will be optimal - at least as far as Kalman theory assumptions are satisfies, which probably means 'not much'.
The code may also make more sense to newcomers, it would be great to have more math people onboard (I wish I did pay more attention at school).

tbolin commented:
I have pushed a new version with a much more Kalman like dynamic gain. There is some trickery to avoid doing more than one  integration, causing the order of updates to be a bit weird compared to a standard Kalman filter.
I'll try to write down a bit more of the math later, but in short linearity is assumed, the noise is assumed to be white and the system is a pure integrator of the gyro rates. 
Works very well on my desk.

I also added a check that prevents updates of the attitude if the time delta is greater than 0.1 sec, and that solved an issue with sudden jumps at start up. The roll/pitch estimate covariance should be reset if this happens, but I haven't implemented that yet.

ledvinap commented:
@tbolin : Great!
Finally, the IMU gain scheduling is  starting to make sense. If it works, it can be adapted to compass and GPS, possibly separating prediction and state update, making it closer to normal implementation.

I need to study the math, but code looks fine.

tbolin commented:
I think I've found the source of the gyro wobble drift:

tbolin commented:
Added settings to adjust the axis scaling. The settings are in the error per full rotation around that axis in decidegrees.

I turned off the acc and did the wobble test with and without adjustmentsm and the drift was was consistently less than half with the adjustments compared to without.

I also made the gyro covariance scale with high gyro rates (still experimental, and should probably be turned down now), and fixed a bug that caused the error vector to always be normalized.

tbolin commented:
Did some flight testing with calibrated gyro rate gains, and I think that helped a lot with the drift during flight. I also had some hard bumps that saturated the gyro and the attitude seems to have recovered quickly, and to a correct value.
[dynamic_gyro_gain_hard_bumps.zip](https://github.com/betaflight/betaflight/files/12198053/dynamic_gyro_gain_hard_bumps.zip)

tbolin commented:
I'm starting to suspect that `imu_gyro_noise_std` and `imu_gyro_noise_std`, isn't standard deviations, but rather noise densities, at least for `imu_gyro_noise_std`.

tbolin commented on src/main/flight/imu.c:
```diff
@@ -247,13 +266,13 @@ STATIC_UNIT_TESTED void imuMahonyAHRSupdate(float dt, float gx, float gy, float
     fpVector3_t mag_bf = {{mag.magADC[X], mag.magADC[Y], mag.magADC[Z]}};
     float recipMagNorm = vectorNormSquared(&mag_bf);
-    if (useMag && recipMagNorm > 0.01f) {
+    if (useMag > 0.0f && recipMagNorm > 0.01f) {
```
Should probably not have been changed, marking so I can get back to it

tbolin commented:
Finally got the tests to compile locally. Now the old tests compile and passes as they should. Tests for the new code is still a todo.

tbolin commented:
rebased

ctzsnooze commented:
@tbolin thanks for all the ongoing work in this PR!

I recall when testing GPS Rescue, that sometimes the Acc would get confused, leading to the quad wanting to fly sideways.  When I aborted the Rescue test, and did a test hover, I'd find that the Acc had a persisting drift that was definitely not present at takeoff.  Confusingly, this new drift problem would persist across arm/disarm cycles.  In contrast, it would be gone if I power cycled - the quad would then hover properly, without drift.

I wondered if the changes here (which are a bit too complicated for my ageing brain to follow) will ensure that while disarmed, the IMU 'resets' to a state like when it first powered up?

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -207,255 +328,298 @@ static float invSqrt(float x)
     if (FLIGHT_MODE(HEADFREE_MODE)) {
-       imuQuaternionComputeProducts(&headfree, &buffer);
+        quaternionProducts buffer;
+        imuQuaternionComputeProducts(&buffer, &headfree);
```
(IIRC all this is for simple rotation around Z)

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -476,7 +640,10 @@ STATIC_UNIT_TESTED float imuCalcCourseErr(float courseOverGround)
     //   over 90 deg, use sign(cross)
     // when well aligned, return ~ 0 (sin(error)), for error >= 90 degrees, return +-1.0
-    return (dot > 0) ? cross : (cross < 0 ? -1.0f : 1.0f);
+    const float scalarErr = ((dot > 0) ? cross : (cross < 0 ? -1.0f : 1.0f) );
```
(technically, this is error vector with Z component only [cross product of vectors in xy plane], not scalar)

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -515,41 +682,36 @@ static void imuDebug_GPS_RESCUE_HEADING(void)
         // project magnetometer reading into Earth frame
         vector3_t mag_ef;
-        matrixVectorMul(&mag_ef, &rMat, &mag_bf); // BF->EF true north
+        matrixVectorMul(&mag_ef, rotMat, mag_bf); // BF->EF true north
```
(wrapper to convert between BF and EF will slightly improve readability)

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -643,55 +807,46 @@ static void imuCalculateEstimatedAttitude(timeUs_t currentTimeUs)
         } else if (gpsSol.groundSpeed > GPS_COG_MIN_GROUNDSPEED) {
             // Reset the reference and reinitialize quaternion factors when GPS groundspeed > GPS_COG_MIN_GROUNDSPEED
-            imuComputeQuaternionFromRPY(&qP, attitude.values.roll, attitude.values.pitch, gpsSol.groundCourse);
+            imuComputeQuaternionFromRPY(&ahrsState.nominal.attitude, attitude.values.roll, attitude.values.pitch, gpsSol.groundCourse);
```
(there are easier ways to align heading)

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -643,55 +807,46 @@ static void imuCalculateEstimatedAttitude(timeUs_t currentTimeUs)
+    imuUpdateEulerAngles(&rMat);
 
-    imuUpdateEulerAngles();
+    attitudeIsEstablished = true;  // TODO conditional on RP covariance
```
(relevant code should probably test covariance directly to see if estimate is good enough)

tbolin replied:
I mostly agree, but I'm not sure what would happen if is isUpright (which uses attitudeIsEstablished) went false during flight. Things would probably be fine, but it will take a lot of checking to be sure.

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -754,12 +912,19 @@ float getCosTiltAngle(void)
+    *state = ahrsState;
+}
+
+void imuGetRotMatrix(matrix33_t* m)
```
(orientationMatrix / attitudeMatrix in future?)

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -778,16 +943,16 @@ void imuSetAttitudeQuat(float w, float x, float y, float z)
-    q.x = x;
-    q.y = y;
-    q.z = z;
+    ahrsState.nominal.attitude.w = w;
```
(or pass quaternion as argument)

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -807,7 +972,12 @@ void imuSetHasNewData(uint32_t dt)
 {
     if ((abs(attitude.values.roll) < 450)  && (abs(attitude.values.pitch) < 450)) {
-        const float yaw = -atan2_approx((+2.0f * (qP.wz + qP.xy)), (+1.0f - 2.0f * (qP.yy + qP.zz)));
+        const quaternion* q = &ahrsState.nominal.attitude;
```
(this should be considerably simpler when using  swing-twist  decomposition to Z axis. But see above :-( )

ledvinap commented on src/main/sensors/gyro.c:
```diff
@@ -531,22 +538,20 @@ FAST_CODE void gyroFiltering(timeUs_t currentTimeUs)
-            gyroFilteredDownsampled[axis] = pt1FilterApply(&gyro.imuGyroFilter[axis], gyro.gyroADCf[axis]);
-        }
-    }
+    // Use clock cycle count at gyro interrupt to get as much precision as possible
```
(gyro is internally using synchronous clock to generate samples. Using constant dt should be more accurate, if no samples are lost - it will remove readout jitter)

tbolin replied:
Assuming a constant dt will effectively create a bias if that dt isn't exactly the average time between calculations. 
There is no derivation, only integration so noise caused by dividing by dt is not an issue.
What is important is that the dt reflects the actual time between measurements. I don't know a more accurate way of doing that than clock cycles between the interrupts from the IMU.

There are also significant jitter in when the attitude task runs, so the number of samples between calculations can vary.

And does all IMUs even have the option to be synchronized to the MCU clock?

ledvinap commented on src/main/sensors/gyro.c:
```diff
@@ -531,22 +538,20 @@ FAST_CODE void gyroFiltering(timeUs_t currentTimeUs)
+    // Use clock cycle count at gyro interrupt to get as much precision as possible
+    // use currentTimeUs converted to cycles if interrupts are not used
+    const bool interruptEnabled = gyro.gyroSensor1.gyroDev.gyroModeSPI != GYRO_EXTI_NO_INT;
+    const uint32_t cycleStamp = interruptEnabled ? gyro.gyroSensor1.gyroDev.gyroLastEXTI : clockMicrosToCycles(currentTimeUs);
```
(with clockMicrosToCycles, data calculation will fail on overflow. Interval between overflows is large, so it is only theoretical problem)

ledvinap commented on src/main/sensors/gyro_init.c:
```diff
@@ -705,6 +700,10 @@ bool gyroInit(void)
         gyro.accSampleRateHz = 0;
     }
 
+    gyro.scalingAdjustment[X] = 360.0f / (360.0f - gyroConfig()->gyro_scaling_adjustment[X] / 10.0f);
```
(all scaling, offset, alignment etc can be factored into 3x4 matrix - like homogeneous coordinates used in graphics)

ledvinap reviewed (commented):
Partial comments, mostly related to existing code quality and not this PR. I'll try to read it again soon.

This PR looks very promising!

tbolin commented on src/main/flight/imu.c:
```diff
@@ -81,31 +82,62 @@ static bool imuUpdated = false;
+static const float IMU_ACC_COVARIANCE_CALC_ACC_NORM_LIMIT = 0.1f;
+// Do not use accelerometer if the gyro norm is greater than this [deg/s]
+static const float IMU_ACC_COVARIANCE_CALC_GYRO_NORM_LIMIT = 50.0f;
+// How fast the gyro covaraince will increase with higher rates
```
agreed

tbolin commented on src/main/flight/imu.c:
```diff
@@ -81,31 +82,62 @@ static bool imuUpdated = false;
+static matrix33_t rMat;
+
+#if defined(UNIT_TEST)
+void imuSetRotMatrix(const matrix33_t* inMatrix) { rMat = *inMatrix; }
```
I agree that it gets messy. I will have a second go at this.

tbolin commented on src/main/flight/imu.c:
```diff
@@ -123,10 +155,86 @@ PG_RESET_TEMPLATE(imuConfig_t, imuConfig,
+void gyroSendTo_imu(const vector3_t* g, const uint32_t stampCycles, const bool overflow)
+{
+    IMU_LOCK;
+    if (imuGyroRecieveState.isInitiated && !overflow) {
```
looks like it. Good catch!

KarateBrot commented on src/main/sensors/gyro.c:
```diff
@@ -531,22 +538,20 @@ FAST_CODE void gyroFiltering(timeUs_t currentTimeUs)
+    const uint32_t cycleStamp = interruptEnabled ? gyro.gyroSensor1.gyroDev.gyroLastEXTI : clockMicrosToCycles(currentTimeUs);
+    const vector3_t gyroVector = {{ gyro.gyroADCf[FD_ROLL], gyro.gyroADCf[FD_PITCH], gyro.gyroADCf[FD_YAW] }};
+
+    extern void gyroSendTo_imu(const vector3_t* g, const uint32_t stampCycles, const bool overflow);
```
I can see a lot of pointer stars attached to the type, not the value. Can you attach it to the value all throughout your PR?
```suggestion
    extern void gyroSendTo_imu(const vector3_t *g, const uint32_t stampCycles, const bool overflow);
```

ctzsnooze commented:
Sorry still haven't tested, but soon will have time.

Could I make one feature request:  could this PR please reset attitude.values.yaw to zero on arming if there is no Mag?  Currently 'heading' is zero when powered up.  But it would be more useful, without a Mag, if this was zeroed each arm, since then if the pilot armed with quad facing due North, which they could remember to do, then the heading of 0 on arm would match the position of the quad immediately after takeoff.

That would help with GPS Rescue or Position Hold initiated soon after takeoff when there is no Mag.

blckmn commented:
Closing this pull request as it has been open for over 12 months without being merged. Thank you for your contribution. If this is still relevant, please feel free to reopen or submit a new pull request.
